### ISSUE TITLE
Host '172.18.0.1' is not allowed to connect to this MySQL server

### ISSUE THREAD

**Issue by RobertCristian** [laradock/laradock#446]
Hello, i really don't know what to do anymore so i'm writing here to get some help. Long story short, i have installed native docker on osx sierra, made a new laravel app, cloned laradock, and started apache2 + mysql + phpmyadmin. I can type localhost and see my app but whenever i try to connect to my database via sequel pro i get a the error from the title and via phpmyadmin, i get  ```connection refused```.

I tried to ssh inside the mysql container and login as root via terminal but i get the following error: 
```ERROR 1045 (28000): Access denied for user 'root'@'localhost' (using password: YES)``` 

I tried to make another clean project, i tried to rebuild with no cache, and i even tried to start only the mysql container without anything, i am really out of ideas and maybe i'm missing something, it's my first docker experience.

Thanks!

**Comment by ELD** [comment]
I'm having the same issue as you. I don't know what the issue is and I've enquired about it in Gitter chat.

**Comment by RobertCristian** [comment]
I just want to add that i've tried on a secondary mac and i get the same error.

**Comment by manshoor** [comment]
@RobertCristian @ELD  have you guys tried `docker volume rm mysql` and rebuild the container?

**Comment by ELD** [comment]
@manshoor Yep, I have removed the volume and rebuilt the container several times. No luck.

**Comment by RobertCristian** [comment]
hm.., when i run `docker volume rm mysql` i get `Error response from daemon: get mysql: no such volume`. But i have tried with `docker-compose down` which deletes everything.

my mysql config from docker-compose.yaml:

``` mysql:
        build:
            context: ./mysql
            args:
                - MYSQL_DATABASE=homestead
                - MYSQL_USER=homestead
                - MYSQL_PASSWORD=secret
                - MYSQL_ROOT_PASSWORD=root
        volumes:
            - mysql:/var/lib/mysql
        ports:
            - "3306:3306"```

**Comment by manshoor** [comment]
can you share your mysql logs? don't add the -d argument and see why mysql is dying.

OR

try with totally refreshing the docker by resetting it to default and pull all the images and build again.

last option would be go back to mysql 5.6

**Comment by RobertCristian** [comment]
```mysql_1               | 2016-11-27T18:50:04.055972Z 0 [Warning] TIMESTAMP with implicit DEFAULT value is deprecated. Please use --explicit_defaults_for_timestamp server option (see documentation for more details).
mysql_1               | 2016-11-27T18:50:04.056781Z 0 [Note] mysqld (mysqld 5.7.16) starting as process 1 ...
mysql_1               | 2016-11-27T18:50:04.059003Z 0 [Note] InnoDB: PUNCH HOLE support available
mysql_1               | 2016-11-27T18:50:04.059052Z 0 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins
mysql_1               | 2016-11-27T18:50:04.059060Z 0 [Note] InnoDB: Uses event mutexes
mysql_1               | 2016-11-27T18:50:04.059066Z 0 [Note] InnoDB: GCC builtin __atomic_thread_fence() is used for memory barrier
mysql_1               | 2016-11-27T18:50:04.059072Z 0 [Note] InnoDB: Compressed tables use zlib 1.2.3
mysql_1               | 2016-11-27T18:50:04.059078Z 0 [Note] InnoDB: Using Linux native AIO
mysql_1               | 2016-11-27T18:50:04.059251Z 0 [Note] InnoDB: Number of pools: 1
mysql_1               | 2016-11-27T18:50:04.059350Z 0 [Note] InnoDB: Using CPU crc32 instructions
mysql_1               | 2016-11-27T18:50:04.060329Z 0 [Note] InnoDB: Initializing buffer pool, total size = 128M, instances = 1, chunk size = 128M
mysql_1               | 2016-11-27T18:50:04.065273Z 0 [Note] InnoDB: Completed initialization of buffer pool
mysql_1               | 2016-11-27T18:50:04.066516Z 0 [Note] InnoDB: If the mysqld execution user is authorized, page cleaner thread priority can be changed. See the man page of setpriority().
mysql_1               | 2016-11-27T18:50:04.078854Z 0 [Note] InnoDB: Highest supported file format is Barracuda.
mysql_1               | 2016-11-27T18:50:04.091509Z 0 [Note] InnoDB: Creating shared tablespace for temporary tables
mysql_1               | 2016-11-27T18:50:04.091654Z 0 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
mysql_1               | 2016-11-27T18:50:04.143611Z 0 [Note] InnoDB: File './ibtmp1' size is now 12 MB.
mysql_1               | 2016-11-27T18:50:04.144295Z 0 [Note] InnoDB: 96 redo rollback segment(s) found. 96 redo rollback segment(s) are active.
mysql_1               | 2016-11-27T18:50:04.144329Z 0 [Note] InnoDB: 32 non-redo rollback segment(s) are active.
mysql_1               | 2016-11-27T18:50:04.144535Z 0 [Note] InnoDB: Waiting for purge to start
mysql_1               | 2016-11-27T18:50:04.197161Z 0 [Note] InnoDB: 5.7.16 started; log sequence number 2533496
mysql_1               | 2016-11-27T18:50:04.197539Z 0 [Note] InnoDB: Loading buffer pool(s) from /var/lib/mysql/ib_buffer_pool
mysql_1               | 2016-11-27T18:50:04.197753Z 0 [Note] Plugin 'FEDERATED' is disabled.
mysql_1               | 2016-11-27T18:50:04.199251Z 0 [Note] InnoDB: Buffer pool(s) load completed at 161127 18:50:04
mysql_1               | 2016-11-27T18:50:04.201533Z 0 [Warning] Failed to set up SSL because of the following SSL library error: SSL context is not usable without certificate and private key
mysql_1               | 2016-11-27T18:50:04.201950Z 0 [Note] Server hostname (bind-address): '0.0.0.0'; port: 3306
mysql_1               | 2016-11-27T18:50:04.201979Z 0 [Note]   - '0.0.0.0' resolves to '0.0.0.0';
mysql_1               | 2016-11-27T18:50:04.202000Z 0 [Note] Server socket created on IP: '0.0.0.0'.
mysql_1               | 2016-11-27T18:50:04.214393Z 0 [Warning] 'user' entry 'root@localhost' ignored in --skip-name-resolve mode.
mysql_1               | 2016-11-27T18:50:04.214454Z 0 [Warning] 'user' entry 'mysql.sys@localhost' ignored in --skip-name-resolve mode.
mysql_1               | 2016-11-27T18:50:04.214477Z 0 [Warning] 'db' entry 'sys mysql.sys@localhost' ignored in --skip-name-resolve mode.
mysql_1               | 2016-11-27T18:50:04.214489Z 0 [Warning] 'proxies_priv' entry '@ root@localhost' ignored in --skip-name-resolve mode.
mysql_1               | 2016-11-27T18:50:04.215027Z 0 [Warning] 'tables_priv' entry 'sys_config mysql.sys@localhost' ignored in --skip-name-resolve mode.
mysql_1               | 2016-11-27T18:50:04.218167Z 0 [Note] Event Scheduler: Loaded 0 events
mysql_1               | 2016-11-27T18:50:04.218284Z 0 [Note] Execution of init_file '/etc/mysql/startup' started.
mysql_1               | 2016-11-27T18:50:04.218741Z 1 [ERROR] 1396  Operation DROP USER failed for 'homestead'@'%'
mysql_1               | 2016-11-27T18:50:04.218850Z 0 [Note] Execution of init_file '/etc/mysql/startup' ended.
mysql_1               | 2016-11-27T18:50:04.218913Z 0 [Note] mysqld: ready for connections.
mysql_1               | Version: '5.7.16'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  MySQL Community Server (GPL)
mysql_1               | 2016-11-27T18:51:40.052632Z 0 [Note] Giving 0 client threads a chance to die gracefully
mysql_1               | 2016-11-27T18:51:40.052702Z 0 [Note] Shutting down slave threads
mysql_1               | 2016-11-27T18:51:40.052718Z 0 [Note] Forcefully disconnecting 0 remaining clients
mysql_1               | 2016-11-27T18:51:40.052733Z 0 [Note] Event Scheduler: Purging the queue. 0 events
mysql_1               | 2016-11-27T18:51:40.052815Z 0 [Note] Binlog end
mysql_1               | 2016-11-27T18:51:40.053365Z 0 [Note] Shutting down plugin 'ngram'
mysql_1               | 2016-11-27T18:51:40.053415Z 0 [Note] Shutting down plugin 'BLACKHOLE'
mysql_1               | 2016-11-27T18:51:40.053429Z 0 [Note] Shutting down plugin 'partition'
mysql_1               | 2016-11-27T18:51:40.053439Z 0 [Note] Shutting down plugin 'ARCHIVE'
mysql_1               | 2016-11-27T18:51:40.053448Z 0 [Note] Shutting down plugin 'MEMORY'
mysql_1               | 2016-11-27T18:51:40.053456Z 0 [Note] Shutting down plugin 'INNODB_SYS_VIRTUAL'
mysql_1               | 2016-11-27T18:51:40.053464Z 0 [Note] Shutting down plugin 'INNODB_SYS_DATAFILES'
mysql_1               | 2016-11-27T18:51:40.053473Z 0 [Note] Shutting down plugin 'INNODB_SYS_TABLESPACES'
mysql_1               | 2016-11-27T18:51:40.053481Z 0 [Note] Shutting down plugin 'INNODB_SYS_FOREIGN_COLS'
mysql_1               | 2016-11-27T18:51:40.053490Z 0 [Note] Shutting down plugin 'INNODB_SYS_FOREIGN'
mysql_1               | 2016-11-27T18:51:40.053529Z 0 [Note] Shutting down plugin 'INNODB_SYS_FIELDS'
mysql_1               | 2016-11-27T18:51:40.053539Z 0 [Note] Shutting down plugin 'INNODB_SYS_COLUMNS'
mysql_1               | 2016-11-27T18:51:40.053548Z 0 [Note] Shutting down plugin 'INNODB_SYS_INDEXES'
mysql_1               | 2016-11-27T18:51:40.053556Z 0 [Note] Shutting down plugin 'INNODB_SYS_TABLESTATS'
mysql_1               | 2016-11-27T18:51:40.053564Z 0 [Note] Shutting down plugin 'INNODB_SYS_TABLES'
mysql_1               | 2016-11-27T18:51:40.053571Z 0 [Note] Shutting down plugin 'INNODB_FT_INDEX_TABLE'
mysql_1               | 2016-11-27T18:51:40.053580Z 0 [Note] Shutting down plugin 'INNODB_FT_INDEX_CACHE'
mysql_1               | 2016-11-27T18:51:40.053587Z 0 [Note] Shutting down plugin 'INNODB_FT_CONFIG'
mysql_1               | 2016-11-27T18:51:40.053595Z 0 [Note] Shutting down plugin 'INNODB_FT_BEING_DELETED'
mysql_1               | 2016-11-27T18:51:40.053603Z 0 [Note] Shutting down plugin 'INNODB_FT_DELETED'
mysql_1               | 2016-11-27T18:51:40.053635Z 0 [Note] Shutting down plugin 'INNODB_FT_DEFAULT_STOPWORD'
mysql_1               | 2016-11-27T18:51:40.053645Z 0 [Note] Shutting down plugin 'INNODB_METRICS'
mysql_1               | 2016-11-27T18:51:40.053653Z 0 [Note] Shutting down plugin 'INNODB_TEMP_TABLE_INFO'
mysql_1               | 2016-11-27T18:51:40.053662Z 0 [Note] Shutting down plugin 'INNODB_BUFFER_POOL_STATS'
mysql_1               | 2016-11-27T18:51:40.053670Z 0 [Note] Shutting down plugin 'INNODB_BUFFER_PAGE_LRU'
mysql_1               | 2016-11-27T18:51:40.053678Z 0 [Note] Shutting down plugin 'INNODB_BUFFER_PAGE'
mysql_1               | 2016-11-27T18:51:40.053686Z 0 [Note] Shutting down plugin 'INNODB_CMP_PER_INDEX_RESET'
mysql_1               | 2016-11-27T18:51:40.053694Z 0 [Note] Shutting down plugin 'INNODB_CMP_PER_INDEX'
mysql_1               | 2016-11-27T18:51:40.053702Z 0 [Note] Shutting down plugin 'INNODB_CMPMEM_RESET'
mysql_1               | 2016-11-27T18:51:40.053709Z 0 [Note] Shutting down plugin 'INNODB_CMPMEM'
mysql_1               | 2016-11-27T18:51:40.053717Z 0 [Note] Shutting down plugin 'INNODB_CMP_RESET'
mysql_1               | 2016-11-27T18:51:40.053751Z 0 [Note] Shutting down plugin 'INNODB_CMP'
mysql_1               | 2016-11-27T18:51:40.053761Z 0 [Note] Shutting down plugin 'INNODB_LOCK_WAITS'
mysql_1               | 2016-11-27T18:51:40.053769Z 0 [Note] Shutting down plugin 'INNODB_LOCKS'
mysql_1               | 2016-11-27T18:51:40.053777Z 0 [Note] Shutting down plugin 'INNODB_TRX'
mysql_1               | 2016-11-27T18:51:40.053785Z 0 [Note] Shutting down plugin 'InnoDB'
mysql_1               | 2016-11-27T18:51:40.053912Z 0 [Note] InnoDB: FTS optimize thread exiting.
mysql_1               | 2016-11-27T18:51:40.054038Z 0 [Note] InnoDB: Starting shutdown...
mysql_1               | 2016-11-27T18:51:40.156959Z 0 [Note] InnoDB: Dumping buffer pool(s) to /var/lib/mysql/ib_buffer_pool
mysql_1               | 2016-11-27T18:51:40.157248Z 0 [Note] InnoDB: Buffer pool(s) dump completed at 161127 18:51:40
mysql_1               | 2016-11-27T18:51:41.303846Z 0 [Note] InnoDB: Shutdown completed; log sequence number 2533524
mysql_1               | 2016-11-27T18:51:41.305578Z 0 [Note] InnoDB: Removed temporary tablespace data file: "ibtmp1"
mysql_1               | 2016-11-27T18:51:41.305604Z 0 [Note] Shutting down plugin 'MyISAM'
mysql_1               | 2016-11-27T18:51:41.305616Z 0 [Note] Shutting down plugin 'MRG_MYISAM'
mysql_1               | 2016-11-27T18:51:41.305622Z 0 [Note] Shutting down plugin 'CSV'
mysql_1               | 2016-11-27T18:51:41.305628Z 0 [Note] Shutting down plugin 'PERFORMANCE_SCHEMA'
mysql_1               | 2016-11-27T18:51:41.305646Z 0 [Note] Shutting down plugin 'sha256_password'
mysql_1               | 2016-11-27T18:51:41.305651Z 0 [Note] Shutting down plugin 'mysql_native_password'
mysql_1               | 2016-11-27T18:51:41.305782Z 0 [Note] Shutting down plugin 'binlog'
mysql_1               | 2016-11-27T18:51:41.306678Z 0 [Note] mysqld: Shutdown complete
mysql_1               |
mysql_1               | 2016-11-27T18:52:56.456211Z 0 [Warning] TIMESTAMP with implicit DEFAULT value is deprecated. Please use --explicit_defaults_for_timestamp server option (see documentation for more details).
mysql_1               | 2016-11-27T18:52:56.457019Z 0 [Note] mysqld (mysqld 5.7.16) starting as process 1 ...
mysql_1               | 2016-11-27T18:52:56.459164Z 0 [Note] InnoDB: PUNCH HOLE support available
mysql_1               | 2016-11-27T18:52:56.459192Z 0 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins
mysql_1               | 2016-11-27T18:52:56.459194Z 0 [Note] InnoDB: Uses event mutexes
mysql_1               | 2016-11-27T18:52:56.459196Z 0 [Note] InnoDB: GCC builtin __atomic_thread_fence() is used for memory barrier
mysql_1               | 2016-11-27T18:52:56.459198Z 0 [Note] InnoDB: Compressed tables use zlib 1.2.3
mysql_1               | 2016-11-27T18:52:56.459199Z 0 [Note] InnoDB: Using Linux native AIO
mysql_1               | 2016-11-27T18:52:56.459435Z 0 [Note] InnoDB: Number of pools: 1
mysql_1               | 2016-11-27T18:52:56.459553Z 0 [Note] InnoDB: Using CPU crc32 instructions
mysql_1               | 2016-11-27T18:52:56.460595Z 0 [Note] InnoDB: Initializing buffer pool, total size = 128M, instances = 1, chunk size = 128M
mysql_1               | 2016-11-27T18:52:56.465638Z 0 [Note] InnoDB: Completed initialization of buffer pool
mysql_1               | 2016-11-27T18:52:56.466835Z 0 [Note] InnoDB: If the mysqld execution user is authorized, page cleaner thread priority can be changed. See the man page of setpriority().
mysql_1               | 2016-11-27T18:52:56.478190Z 0 [Note] InnoDB: Highest supported file format is Barracuda.
mysql_1               | 2016-11-27T18:52:56.494552Z 0 [Note] InnoDB: Creating shared tablespace for temporary tables
mysql_1               | 2016-11-27T18:52:56.494703Z 0 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
mysql_1               | 2016-11-27T18:52:56.546024Z 0 [Note] InnoDB: File './ibtmp1' size is now 12 MB.
mysql_1               | 2016-11-27T18:52:56.546674Z 0 [Note] InnoDB: 96 redo rollback segment(s) found. 96 redo rollback segment(s) are active.
mysql_1               | 2016-11-27T18:52:56.546693Z 0 [Note] InnoDB: 32 non-redo rollback segment(s) are active.
mysql_1               | 2016-11-27T18:52:56.546924Z 0 [Note] InnoDB: Waiting for purge to start
mysql_1               | 2016-11-27T18:52:56.599523Z 0 [Note] InnoDB: 5.7.16 started; log sequence number 2533524
mysql_1               | 2016-11-27T18:52:56.600039Z 0 [Note] InnoDB: Loading buffer pool(s) from /var/lib/mysql/ib_buffer_pool
mysql_1               | 2016-11-27T18:52:56.600361Z 0 [Note] Plugin 'FEDERATED' is disabled.
mysql_1               | 2016-11-27T18:52:56.602230Z 0 [Note] InnoDB: Buffer pool(s) load completed at 161127 18:52:56
mysql_1               | 2016-11-27T18:52:56.603445Z 0 [Warning] Failed to set up SSL because of the following SSL library error: SSL context is not usable without certificate and private key
mysql_1               | 2016-11-27T18:52:56.603992Z 0 [Note] Server hostname (bind-address): '0.0.0.0'; port: 3306
mysql_1               | 2016-11-27T18:52:56.604018Z 0 [Note]   - '0.0.0.0' resolves to '0.0.0.0';
mysql_1               | 2016-11-27T18:52:56.604091Z 0 [Note] Server socket created on IP: '0.0.0.0'.
mysql_1               | 2016-11-27T18:52:56.619428Z 0 [Warning] 'user' entry 'root@localhost' ignored in --skip-name-resolve mode.
mysql_1               | 2016-11-27T18:52:56.619474Z 0 [Warning] 'user' entry 'mysql.sys@localhost' ignored in --skip-name-resolve mode.
mysql_1               | 2016-11-27T18:52:56.619497Z 0 [Warning] 'db' entry 'sys mysql.sys@localhost' ignored in --skip-name-resolve mode.
mysql_1               | 2016-11-27T18:52:56.619506Z 0 [Warning] 'proxies_priv' entry '@ root@localhost' ignored in --skip-name-resolve mode.
mysql_1               | 2016-11-27T18:52:56.620224Z 0 [Warning] 'tables_priv' entry 'sys_config mysql.sys@localhost' ignored in --skip-name-resolve mode.
mysql_1               | 2016-11-27T18:52:56.623879Z 0 [Note] Event Scheduler: Loaded 0 events
mysql_1               | 2016-11-27T18:52:56.624051Z 0 [Note] Execution of init_file '/etc/mysql/startup' started.
mysql_1               | 2016-11-27T18:52:56.624569Z 1 [ERROR] 1396  Operation DROP USER failed for 'homestead'@'%'
mysql_1               | 2016-11-27T18:52:56.624724Z 0 [Note] Execution of init_file '/etc/mysql/startup' ended.
mysql_1               | 2016-11-27T18:52:56.624811Z 0 [Note] mysqld: ready for connections.
mysql_1               | Version: '5.7.16'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  MySQL Community Server (GPL)```

**Comment by manshoor** [comment]
ok, so what is the `docker ps ` says because according to this `[Note] mysqld: ready for connections`

in mysql container try `mysql -u root -p root`

**Comment by RobertCristian** [comment]
Same as in my first post: `ERROR 1045 (28000): Access denied for user 'root'@'localhost' (using password: YES)`

**Comment by ELD** [comment]
So it looks like the MySQL root user can't be accessed with a password. If you're inside the MySQL container, you need to login as root without a password (`mysql -u root`). Once there, you can manually create the Homestead database and the user accordingly.

I think the issue has something to do with the startup script. It fails on dropping the user and if you modify the command to use MySQL 5.7's `DROP USER IF EXISTS` syntax, it fails on creating the homestead user. I'm not sure what the cause is. I'm playing around with it to troubleshoot the creation of the image so that it doesn't crash on the first startup. It seems that the startup script runs fine the second time you start the image, though.

**Comment by ELD** [comment]
Just to further document what I'm looking at, this line jumps out at me in the logs for the mysql container:
```
mysql_1               | 2016-11-27T20:46:21.614859Z 1 [Warning] root@localhost is created with an empty password ! Please consider switching off the --initialize-insecure option.
```
Looks like that's why we can't log into the root user with the password. It wasn't set up with one.

Secondly, it looks like the `startup` script is failing:
```
mysql_1               | ERROR 1396 (HY000) at line 1: Operation CREATE USER failed for 'homestead'@'%'
```

Additionally, after building the container, this is what `docker-compose ps` displays:

```
      Name                    Command               State    Ports
------------------------------------------------------------------
laradock_mysql_1   docker-entrypoint.sh mysql ...   Exit 1
```

**Comment by kamilcios** [comment]
I had the same problem, try removing first line from mysql/startup ( DROP USER 'homestead';
 ). Worked for me.

**Comment by ELD** [comment]
@kamilcios That doesn't work for me. After building the image and starting up the container, the first time it still exits prematurely. Instead of failing on dropping the user, it fails on trying to create the user.

From looking into the history, I think the culprit is #437 and #438. I just can't figure out the chronology because one reverts the other but the master branch doesn't reflect a reversion. I don't understand it.

**Comment by philtrep** [comment]
@kamilcios Could you do a PR for your fix?

@ELD I merged it by mistake the first-time so I reverted it ASAP, I then had discussion with @Mahmoudz then merged it back in. My apologies for the confusion, the mobile github site is the bane of my existence.

This seems to be creating many problems, I could remove the PR but I think it's worth trying to keep the functionality of creating custom users (if possible). That being said if a majority doesn't care about the feature, I'll definitely remove it.

**Comment by ELD** [comment]
@philtrep Okay, that makes sense. Yeah, it looks like a better way to go about creating users, it's just that the startup script may error out and I think when that happens, the entire MySQL server dies. I'm not positive though.

Definitely worth trying to make it work better.

**Comment by Blaasvis** [comment]
This works for me. It didn't crash on the first create for me anymore which it did before. 
Now phpmyadmin is complaining on first boot about not getting addrinfo. But second run it works fine. 

`DROP USER IF EXISTS 'MYSQL_USER';
CREATE USER 'MYSQL_USER'@'%';
CREATE DATABASE IF NOT EXISTS MYSQL_DATABASE;
GRANT ALL ON `MYSQL_DATABASE`.* TO 'MYSQL_USER'@'%' IDENTIFIED BY 'MYSQL_PASSWORD';`

**Comment by Blaasvis** [comment]
created the pull request just now.

**Comment by philtrep** [comment]
@Blaasvis 's PR is #451. I'll wait a few hours for people to review it before merging.

**Comment by philtrep** [comment]
Resolved by #451
